src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -262,6 +262,58 @@ FabricInfo * RetrieveCurrentFabric(CommandHandler * aCommandHandler)
     return Server::GetInstance().GetFabricTable().FindFabricWithIndex(index);
 }
 
+void FailSafeCleanup(FailSafeContext & failSafeContext)
+{
+    FabricInfo * fabricInfo = Server::GetInstance().GetFabricTable().FindFabricWithIndex(failSafeContext.GetFabricIndex());
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Call to FailSafeCleanup");
+
+    VerifyOrReturn(fabricInfo != nullptr);
+
+    // If an AddNOC or UpdateNOC command has been successfully invoked, terminate all CASE sessions associated with the Fabric
+    // whose Fabric Index is recorded in the Fail-Safe context (see ArmFailSafe Command) by clearing any associated Secure
+    // Session Context at the Server.
+    if (failSafeContext.NocCommandHasBeenInvoked())
```

Comment:
As far as I can tell, the order of things now is:

1. Call `PostEvent` in `CommissioningFailedTimerComplete`.
2. Set `mAddNocCommandHasBeenInvoked` and `mUpdateNocCommandHasBeenInvoked` to false.
3. Land in this code (because the event gets handled async).

So as far as I can tell this will always test false.

What am I missing?  If I am not missing anything, why did the testing that was done show this working?

Also if I am not missing anything that means we need to capture the two "command has been invoked" booleans in the event.  Those and the fabric index are the only things this function uses from the failsafe context, and the index is already in the event.

And we should have tests here.  Doing AddNOC in yaml is not exactly trivial right now, but if we can add a way to do that, we could presumably test this with a very short failsafe period.....

Reply:
Acked, my test has very short failsafe period and CommissioningFailedTimerComplete get fired before AddNoC/UpdateNoc,  trying to add yaml, but it seems not trivial task

Reply:
I'm fine with the tests as a followup, fwiw, as long as we have some manual testing that makes sure we actually enter the cleanup code correctly.

Reply:
https://github.com/project-chip/connectedhomeip/issues/16379